File: Matrix_Chain_Multiplication/Matrix_Chain_Multiplication_Dp.py
def matrix_product(p):
    """Return m and s.
 
    m[i][j] is the minimum number of scalar multiplications needed to compute the
    product of matrices A(i), A(i + 1), ..., A(j).
 
    s[i][j] is the index of the matrix after which the product is split in an
    optimal parenthesization of the matrix product.
 
    p[0... n] is a list such that matrix A(i) has dimensions p[i - 1] x p[i].
    """
    length = len(p) # len(p) = number of matrices + 1
 
    # m[i][j] is the minimum number of multiplications needed to compute the
    # product of matrices A(i), A(i+1), ..., A(j)
    # s[i][j] is a matrix after which the product is split in the minimum
    # number of multiplications needed
    m = [[-1]*length for _ in range(length)]
    s = [[-1]*length for _ in range(length)]
 
    matrix_product_helper(p, 1, length - 1, m, s)
 
    return m, s
 
 
def matrix_product_helper(p, start, end, m, s):
    """Return minimum number of scalar multiplications needed to compute the
    product of matrices A(start), A(start + 1), ..., A(end).
 
    The minimum number of scalar multiplications needed to compute the
    product of matrices A(i), A(i + 1), ..., A(j) is stored in m[i][j].
 
    The index of the matrix after which the above product is split in an optimal
    parenthesization is stored in s[i][j].
 
    p[0... n] is a list such that matrix A(i) has dimensions p[i - 1] x p[i].
    """
    if m[start][end] >= 0:
        return m[start][end]
 
    if start == end:
        q = 0
    else:
        q = float('inf')
        for k in range(start, end):
            temp = matrix_product_helper(p, start, k, m, s) \
                   + matrix_product_helper(p, k + 1, end, m, s) \
                   + p[start - 1]*p[k]*p[end]
            if q > temp:
                q = temp
                s[start][end] = k
 
    m[start][end] = q
    return q
 
 
def print_parenthesization(s, start, end):
    """Print the optimal parenthesization of the matrix product A(start) x
    A(start + 1) x ... x A(end).
 
    s[i][j] is the index of the matrix after which the product is split in an
    optimal parenthesization of the matrix product.
    """
    if start == end:
        print('A[{}]'.format(start), end='')
        return
 
    k = s[start][end]
 
    print('(', end='')
    print_parenthesization(s, start, k)
    print_parenthesization(s, k + 1, end)
    print(')', end='')
 
 
n = int(input('Enter number of matrices: '))
p = []
for i in range(n):
    temp = int(input('Enter number of rows in matrix {}: '.format(i + 1)))
    p.append(temp)
temp = int(input('Enter number of columns in matrix {}: '.format(n)))
p.append(temp)
 
m, s = matrix_product(p)
print('The number of scalar multiplications needed:', m[1][n])
print('Optimal parenthesization: ', end='')
print_parenthesization(s, 1, n)


#Output- 
 
# Case 1:
# Enter number of matrices: 3
# Enter number of rows in matrix 1: 10
# Enter number of rows in matrix 2: 100
# Enter number of rows in matrix 3: 5
# Enter number of columns in matrix 3: 50
# The number of scalar multiplications needed: 7500
# Optimal parenthesization: ((A[1]A[2])A[3])
 
# Case 2:
# Enter number of matrices: 5
# Enter number of rows in matrix 1: 5
# Enter number of rows in matrix 2: 10
# Enter number of rows in matrix 3: 8
# Enter number of rows in matrix 4: 15
# Enter number of rows in matrix 5: 20
# Enter number of columns in matrix 5: 4
# The number of scalar multiplications needed: 2200
# Optimal parenthesization: (A[1](A[2](A[3](A[4]A[5]))))
 

File: Matrix_Chain_Multiplication/Matrix_Chain_Multiplication_DP.java
/*
 * A Dynamic Programming approach to find the minimum(optimal) number of operations required to multiply a chain of matrices
 * Time complexity = O(n^3)
 * Space complexity = O(n^2)
 */

import java.util.Scanner;

public class matrixChainMultiplication { 
	
	//function to find the optimized number of operations required for multiplication using DP
	static int orderOfMultiplication(int arr[]) { 
		int size = arr.length;
		
		//2D array to store optimal solution for all i
		int dp[][] = new int[size][size]; 
		
		// cost will be zero for multiplying one matrix
		for (int i=1; i<size; i++) { 
			dp[i][i] = 0; 
		}
		
		// l is the chain length
		for (int l=2; l<size; l++) { 
			for (int i=1; i<size-l+1; i++){ 
				int j = i+l-1; 
				if(j == size) {
					continue; 
				}
				dp[i][j] = Integer.MAX_VALUE; 
				for (int k=i; k<=j-1; k++) { 
					// temp is cost/scalar multiplications 
					int temp = dp[i][k] + dp[k+1][j] + (arr[i-1] * arr[k] * arr[j]); 
					
					//if this order gives better solution
					if (temp < dp[i][j]) {
						//update dp[i][j]
						dp[i][j] = temp; 
					}
				} 
			} 
		} 
		//optimal solution for the chain multiplication will be at the dp[1][size-1]
		return dp[1][size-1]; 
	} 

	//Main function
	public static void main(String args[]) { 
		Scanner s = new Scanner(System.in);
		System.out.print("Enter size of array - ");
		int size = s.nextInt();
		
		//storing matrices sizes given by user
		int arr[] = new int[size];
		System.out.print("Enter values of array - ");
		for(int i=0; i<size; i++) {
			arr[i] = s.nextInt();
		}
		
		//function calling
		int ans = orderOfMultiplication(arr); 
		System.out.println("Minimum number of operations required - " + ans);
		s.close();
	} 
} 

/*
 * =========================================================
 * Sample Input-Output
 * =========================================================
 * Input :  Enter size of array - 4
 * 			Enter values of array - 2 4 6 8
 * Output : Minimum number of operations required - 144
 * =========================================================
 */
